### PR TITLE
Cleanup the various nbgrader apps

### LINKED ISSUE #57: Reorganize the various nbgrader apps
They're currently a bit disorganized, and there's some code duplication that I'm unhappy about (e.g. overriding the `ipython_dir` trait). This is not a major issue, but should be addressed at some point.

### PR DESCRIPTION
**Pull request by jhamrick**
Work in progress to clean up the apps, make there be less code redundancy, and make the interface a little more intuitive and less complicated. This will fix #57 when complete, and address at least part of #66

### CODE REVIEW

File: nbgrader/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from exporters import AssignmentExporter
```
Comment (Carreau) [inline nbgrader/__init__.py]: Travis seem sad about that. 

Maybe relative import ? `.exporter` ?

**Comment by jhamrick** [comment]
Travis is failing currently because this depends on some new commits in IPython that aren't in IPython 3 yet.

**Comment by Carreau** [comment]
> Travis is failing currently because this depends on some new commits in IPython that aren't in IPython 3.

Just pip install -e from master/3.x :-)

**Comment by jhamrick** [comment]
Woohoo, I am done with this I think! This adds a bunch of tests and documentation and in general makes things a lot cleaner. Importantly:

* all the nbgrader apps now either inherit from a `BaseNbGraderApp` (which sets various defaults, like the ipython profile to be 'nbgrader'), or `BaseNbConvertApp` (which inherits from both `NbConvertApp` in IPython and `BaseNbGraderApp`)
* there is now an `AssignmentExporter` that is used by the apps that inherit from `BaseNbConvertApp`, which makes sure to include assignment-relevant things in the resources dict, including the assignment id, student id, and information about the database.
* the `--recursive` option has been removed, because I think it was only really important for `nbgrader autograde`, but in practice it's easier to just use `nbgrader autograde` on one submission at a time, and write scaffolding code around it to loop over all the students/submissions.

This significantly reduces the amount of code duplication, make nbgrader easier to understand, and also more robust, I think. I will merge this once https://github.com/ipython/ipython/pull/7912 is merged (the tests here will fail until that one is merged).

Ping @ellisonbg

**Comment by ellisonbg** [comment]
Jess, I won't have time to look at this before next week, so feel free to
merge this and iterate in master as needed.

On Mon, Mar 2, 2015 at 5:38 PM, Jessica B. Hamrick <notifications@github.com
> wrote:

> Woohoo, I am done with this I think! This adds a bunch of tests and
> documentation and in general makes things a lot cleaner. Importantly:
>
>    - all the nbgrader apps now either inherit from a BaseNbGraderApp
>    (which sets various defaults, like the ipython profile to be 'nbgrader'),
>    or BaseNbConvertApp (which inherits from both NbConvertApp in IPython
>    and BaseNbGraderApp)
>    - there is now an AssignmentExporter that is used by the apps that
>    inherit from BaseNbConvertApp, which makes sure to include
>    assignment-relevant things in the resources dict, including the assignment
>    id, student id, and information about the database.
>    - the --recursive option has been removed, because I think it was only
>    really important for nbgrader autograde, but in practice it's easier
>    to just use nbgrader autograde on one submission at a time, and write
>    scaffolding code around it to loop over all the students/submissions.
>
> This significantly reduces the amount of code duplication, make nbgrader
> easier to understand, and also more robust, I think. I will merge this once
> ipython/ipython#7912 <https://github.com/ipython/ipython/pull/7912> is
> merged (the tests here will fail until that one is merged).
>
> Ping @ellisonbg <https://github.com/ellisonbg>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/nbgrader/pull/112#issuecomment-76870130>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by jhamrick** [comment]
Ok, thanks!

**Pull request merged**